### PR TITLE
Revamp APIs around bottle specifications

### LINKED ISSUE #12068: `all` bottles are built from source despite being `cellar: :any`
### `brew config` output

```shell
HOMEBREW_VERSION: 3.2.12-18-g2dd3934
ORIGIN: https://github.com/Homebrew/brew
HEAD: 2dd3934fd791dbf5676a6113a0dea12066422b5b
Last commit: 3 hours ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 844c6ae669bd7b7d7000a0c8b2243f9d30404bf2
Core tap last commit: 6 minutes ago
Core tap branch: master
HOMEBREW_PREFIX: /Users/carlocab/homebrew
HOMEBREW_REPOSITORY: /Users/carlocab/homebrew
HOMEBREW_CELLAR: /Users/carlocab/homebrew/Cellar
HOMEBREW_CASK_OPTS: []
HOMEBREW_COLOR: set
HOMEBREW_EDITOR: nvim
HOMEBREW_GITHUB_PACKAGES_TOKEN: set
HOMEBREW_GITHUB_PACKAGES_USER: carlocab
HOMEBREW_GIT_EMAIL: 30379873+carlocab@users.noreply.github.com
HOMEBREW_GIT_NAME: Carlo Cabrera
HOMEBREW_MAKE_JOBS: 4
HOMEBREW_NO_AUTO_UPDATE: set
HOMEBREW_NO_BOOTSNAP: set
HOMEBREW_PRY: set
Homebrew Ruby: 2.6.3 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: quad-core 64-bit icelake
Clang: 12.0.5 build 1205
Git: 2.30.1 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 7.64.1 => /usr/bin/curl
macOS: 11.5.2-x86_64
CLT: 12.5.0.22.9
Xcode: N/A
```

### `brew doctor` output

```shell
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: "config" scripts exist outside your system or Homebrew directories.
`./configure` scripts often look for *-config scripts to determine if
software packages are installed, and which additional flags to use when
compiling and linking.

Having additional scripts in your path can confuse software installed via
Homebrew if the config script overrides a system or Homebrew-provided
script of the same name. We found the following "config" scripts:
  /usr/local/bin/pkg-config
  /usr/local/bin/jemalloc-config
  /usr/local/bin/freetype-config
  /usr/local/bin/libpng16-config
  /usr/local/bin/libpng-config
  /usr/local/bin/gpgrt-config
  /usr/local/bin/python3.9-config
  /usr/local/bin/onig-config
  /usr/local/bin/ksba-config
  /usr/local/bin/gsl-config
  /usr/local/bin/python3-config
  /usr/local/bin/npth-config
  /usr/local/bin/libassuan-config
  /usr/local/bin/pcre2-config
  /usr/local/bin/gpg-error-config
  /usr/local/bin/pcre-config
  /usr/local/bin/gts-config
  /usr/local/bin/gpgme-config
  /usr/local/bin/libgcrypt-config
  /usr/local/bin/sdl2-config
  /usr/local/bin/guile-config

Warning: gettext files detected at a system prefix.
These files can cause compilation and link failures, especially if they
are compiled with improper architectures. Consider removing these files:
  /Users/carlocab/homebrew/lib/libgettextlib.dylib
  /Users/carlocab/homebrew/lib/libintl.dylib
  /Users/carlocab/homebrew/include/libintl.h
  /usr/local/lib/libgettextlib.dylib
  /usr/local/lib/libintl.dylib
  /usr/local/include/libintl.h

Warning: Your Homebrew's prefix is not /usr/local.
Some of Homebrew's bottles (binary packages) can only be used with the default
prefix (/usr/local).
You will encounter build failures with some formulae.
Please create pull requests instead of asking for help on Homebrew's GitHub,
Twitter or any other official channels. You are responsible for resolving
any issues you experience while you are running this
unsupported configuration.


Warning: /usr/bin occurs before /Users/carlocab/homebrew/bin in your PATH.
This means that system-provided programs will be used instead of those
provided by Homebrew. Consider setting your PATH so that
/Users/carlocab/homebrew/bin occurs before /usr/bin. Here is a one-liner:
  echo 'export PATH="/Users/carlocab/homebrew/bin:$PATH"' >> ~/.zshrc

The following tools exist at both paths:
  python3
  pip3

Warning: Homebrew's "bin" was not found in your PATH.
Consider setting your PATH for example like so:
  echo 'export PATH="/Users/carlocab/homebrew/bin:$PATH"' >> ~/.zshrc
```

### 

- [X] I ran `brew update` and am still able to reproduce my issue.
- [X] I have resolved all warnings from `brew doctor` and that did not fix my problem.

### What were you trying to do (and why)?

Install `eigen` from a bottle.

### What happened (include all command output)?

It built from source:
```
❯ devbrew install eigen
==> Downloading https://ghcr.io/v2/homebrew/core/cmake/manifests/3.21.2
Already downloaded: /Users/carlocab/Library/Caches/Homebrew/downloads/e53caefe5de8095f304f2da999d9a73844e1359e599235adc9ab82512ffe4b1a--cmake-3.21.2.bottle_manifest.json
==> Downloading https://ghcr.io/v2/homebrew/core/cmake/blobs/sha256:dbece5192774308bf40ee7df06fdd4bba9ee6671e4e2e740072ec04cb9bef0c2
Already downloaded: /Users/carlocab/Library/Caches/Homebrew/downloads/1a610e91c4b40f0ebf65f8d76389e773d2b7cf8891a5175f390a0682c4d30617--cmake--3.21.2.big_sur.bottle.tar.gz
Warning: Building eigen from source as the bottle needs:
- HOMEBREW_CELLAR: /usr/local/Cellar (yours is /Users/carlocab/homebrew/Cellar)
- HOMEBREW_PREFIX: /usr/local (yours is /Users/carlocab/homebrew)
==> Downloading https://gitlab.com/libeigen/eigen/-/archive/3.4.0/eigen-3.4.0.tar.gz
          -#O=-#     #    #
==> Installing dependencies for eigen: cmake
==> Installing eigen dependency: cmake
==> Pouring cmake--3.21.2.big_sur.bottle.tar.gz
🍺  /Users/carlocab/homebrew/Cellar/cmake/3.21.2: 6,615 files, 69MB
==> Installing eigen
==> cmake -Dpkg_config_libdir=/Users/carlocab/homebrew/Cellar/eigen/3.4.0_1/lib ..
==> make install
🍺  /Users/carlocab/homebrew/Cellar/eigen/3.4.0_1: 546 files, 8.4MB, built in 31 seconds
```

### What did you expect to happen?

`eigen` is `cellar: :any_skip_relocation`, so `brew` should've poured a bottle.

### Step-by-step reproduction instructions (by running `brew` commands)

```shell
cd $HOME
mkdir homebrew && curl -L https://github.com/Homebrew/brew/tarball/master | tar xz --strip 1 -C homebrew
~/homebrew/bin/brew install eigen
```

### PR DESCRIPTION
**Pull request by Bo98**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

Goals:

* Stronger typing (rather than passing around hashes everywhere)
* Less confusing names (e.g. `checksum_for` returning cellar information)
* Remove relics from old bottle spec DSL.

Additions:

* `BottleSpecification#tag_specification_for`
  - For a given tag, returns a new `TagSpecification` object containing the actual tag (since it may choose an older OS), the checksum and the cellar value.
* `Utils::Bottles::Collector#specification_for`
  - Backend for the above.
* Add `tag` parameter to `BottleSpecification#compatible_locations?` (default: tag for user's OS)
  - `Bottle#compatible_locations?` now uses this internally.
* Add `tag` parameter to `BottleSpecification#skip_relocation?` (default: tag for user's OS)
  - `Bottle#skip_relocation?` now uses this internally.
* `Utils::Bottles::Collector#tags`
  - Returns tags present in the collector
* `Utils::Bottles::Collector#each_tag`
  - For enumeration over the tags in the collector
* `Utils::Bottles::Collector#tag?`
  - Behaves just like `BottleSpecification#tag?` did (same parameters). Now a backend for that.
* `Utils::Bottles::Collector#add`
  - Pass in a tag, the checksum and the cellar and it'll create and store a `TagSpecification` object into the collector

Removals:

* `BottleSpecification#cellar`
  - Replacement: Use `BottleSpecification#tag_specification_for` and get the cellar value from that.
* `BottleSpecification#all_tags_cellar`
  - Always returned nil.
* `BottleSpecification#prefix`
  - Always returned `Homebrew::DEFAULT_PREFIX`.
* `Bottle#prefix`
  - As above.
* `BottleSpecification#checksum_for`
  - Replacement: `BottleSpecification#tag_specification_for`
* `Utils::Bottles::Collector#keys`
  - Replacement: `Utils::Bottles::Collector#tags` (returns `Util::Bottles::Tag`s rather than symbols)
* `Utils::Bottles::Collector#each_key`
  - Replacement: `Utils::Bottles::Collector#each_tag` (passes `Util::Bottles::Tag`s rather than symbols)
* `Utils::Bottles::Collector#key?`
  - Replacement: `Utils::Bottles::Collector#tag?` (takes a `Util::Bottles::Tag` rather than a symbol, and now also looks back for older OS versions or `:all` bottles like `BottleSpecification#tag?` did - if this is not desirable then search `Utils::Bottles::Collector#tags`)
* `Utils::Bottles::Collector#[]`
  - Replacement: `Utils::Bottles::Collector#specification_for` (returns `TagSpecifiation` object rather than a hash, and also looks back for older OS versions or `:all` bottles - if this is not desirable then search first if the tag exists in `Utils::Bottles::Collector#tags`)
* `Utils::Bottles::Collector#[]=`
  - Replacement: `Utils::Bottles::Collector#add` (pass in checksum and cellar as parameters rather than just passing a hash)
* `Utils::Bottles::Collector#dig`
  - Replacement: `Utils::Bottles::Collector#specification_for` (see `Utils::Bottles::Collector#[]` deprecation)

Newer APIs more strictly require `Utils::Bottles::Tag` rather than also allowing symbols.

The changes here hopefully fix #12068, but I haven't specifically tested that scenario yet.

### CODE REVIEW

**Comment by Bo98** [comment]
`Utils::Bottles::Collector` stuff could maybe be removed completely rather than deprecated. `Utils::Bottles::Collector` is technically private API, but the `BottleSpecification#collector` getter is public API.

File: Library/Homebrew/dev-cmd/bottle.rb [hunk 1]
```diff
@@ -551,7 +551,7 @@ def bottle_formula(f, args:)
         },
         "bottle"  => {
           "root_url" => bottle.root_url,
-          "prefix"   => bottle.prefix,
+          "prefix"   => Homebrew::DEFAULT_PREFIX, # TODO: deprecate or parse this from cellar value
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/bottle.rb:554]: Is this correct if you're bottling in a non-default prefix/cellar? I agree it should be ideal deprecated, though. Could you chuck a `3.3.0` reference in that comment somewhere?
```suggestion
          "prefix"   => Homebrew::DEFAULT_PREFIX, # TODO: 3.3.0 deprecate this
```

File: Library/Homebrew/formula_installer.rb [hunk 2]
```diff
@@ -156,7 +156,6 @@ def pour_bottle?(output_warning: false)
         opoo <<~EOS
           Building #{formula.full_name} from source as the bottle needs:
           - HOMEBREW_CELLAR: #{bottle.cellar} (yours is #{HOMEBREW_CELLAR})
-          - HOMEBREW_PREFIX: #{bottle.prefix} (yours is #{HOMEBREW_PREFIX})
```
Comment (MikeMcQuaid) [inline Library/Homebrew/formula_installer.rb:159]: I think the prefix being correct is needed, no?

File: Library/Homebrew/software_spec.rb [hunk 3]
```diff
@@ -314,19 +314,24 @@ def initialize(formula, spec, tag = nil)
...
 
     @resource.version = formula.pkg_version
-    @resource.checksum = checksum
+    @resource.checksum = tag_spec.checksum
 
     root_url(spec.root_url, spec.root_url_specs)
   end
 
+  def prefix
+    # odeprecated "Bottle#prefix"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/software_spec.rb:330]:
```suggestion
    # TODO: 3.3.0: odeprecated "Bottle#prefix"
```

File: Library/Homebrew/software_spec.rb [hunk 4]
```diff
@@ -505,34 +509,52 @@ def cellar(val = nil)
...
+    nil
   end
 
-  def compatible_locations?
-    # this looks like it should check prefix and repository too but to be
-    # `cellar :any` actually requires no references to the cellar, prefix or
-    # repository.
-    return true if [:any, :any_skip_relocation].include?(cellar)
+  def prefix
+    # odeprecated "BottleSpecification#prefix"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/software_spec.rb:529]:
```suggestion
    # TODO: 3.3.0: odeprecated "BottleSpecification#prefix"
```

File: Library/Homebrew/software_spec.rb [hunk 5]
```diff
@@ -584,25 +613,27 @@ def sha256(hash)
     )
   }
   def checksum_for(tag, no_older_versions: false)
+    # odeprecated "BottleSpecification#checksum_for", "BottleSpecification#tag_specification_for"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/software_spec.rb:616]:
```suggestion
    # TODO: 3.3.0: odeprecated "BottleSpecification#checksum_for", "BottleSpecification#tag_specification_for"
```

File: Library/Homebrew/utils/bottles.rb [hunk 6]
```diff
@@ -217,17 +225,112 @@ def default_cellar
...
+      extend T::Sig
 
       sig { void }
       def initialize
-        @checksums = {}
+        @tag_specs = T.let({}, T::Hash[Utils::Bottles::Tag, Utils::Bottles::TagSpecification])
+      end
+
+      def keys
+        # odeprecated "Utils::Bottles::Collector#keys", "Utils::Bottles::Collector#tags"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/bottles.rb:258]:
```suggestion
        # odeprecated: 3.3.0: "Utils::Bottles::Collector#keys", "Utils::Bottles::Collector#tags"
```

File: Library/Homebrew/utils/bottles.rb [hunk 7]
```diff
@@ -217,17 +225,112 @@ def default_cellar
...
+        tags.map(&:to_sym)
+      end
+
+      sig { returns(T::Array[Utils::Bottles::Tag]) }
+      def tags
+        @tag_specs.keys
+      end
+
+      def [](tag)
+        # odeprecated "Utils::Bottles::Collector#[]", "Utils::Bottles::Collector#specification_for"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/bottles.rb:269]:
```suggestion
        # odeprecated: 3.3.0: "Utils::Bottles::Collector#[]", "Utils::Bottles::Collector#specification_for"
```

**Review by MikeMcQuaid — APPROVED** [review]
Looks good so far. Can any of these deprecated methods be just removed now? These aren't public APIs. Will this fix the previously mentioned `:all` issues?

**Comment by MikeMcQuaid** [comment]
> Just realised the `Utils::Bottles::Collector` stuff could maybe be removed completely rather than deprecated. `Utils::Bottles` is technically private API (even `Utils::Bottles::Tag`, which it probably shouldn't be)

Do it 👍🏻

File: Library/Homebrew/formula_installer.rb (hunk 2 above)
Comment (Bo98) [reply Library/Homebrew/formula_installer.rb:159]: `Bottle#prefix` is deprecated and always returns the default prefix.

Comment (MikeMcQuaid) [reply Library/Homebrew/formula_installer.rb:159]: Gotcha (I did that, oops). I think I saw a user complaint about that somewhere but cannot find it. Probably need to just bake in the assumption that prefix is `Cellar/..`?

Comment (Bo98) [reply Library/Homebrew/formula_installer.rb:159]: Oh for the `HOMEBREW_REPOSITORY/Cellar` case? I'll try do something.

Comment (MikeMcQuaid) [reply Library/Homebrew/formula_installer.rb:159]: Yeh, something like that.

**Review by MikeMcQuaid — APPROVED** [review]
Looks great. I think all methods not explicitly documented as public can be removed. This is very much "Homebrew internals" territory and we're very far from "stuff a Formula would reasonable use"

**Comment by Bo98** [comment]
> not explicitly documented as public

Depends what you mean by this. We almost never explicitly write `@api public`, and when we do it's because we're overriding a previous `@api private`.

I've removed `Utils::Bottles::Collector` methods since they were explicitly `@api private`.

I do however see that `Formula#bottle` is private so in theory `Bottle#prefix` could be removed since there's no public getter for a bottle object (unless you create the object yourself). Nothing explicit is written about `BottleSpecification`, though the getter does appear to be missing from https://rubydoc.brew.sh/Formula (we don't mark it as private so this is probably actually a doc generation error but I suppose that means it's been private all along).

**Comment by MikeMcQuaid** [comment]
> `Bottle#prefix` could be removed since there's no public getter for a bottle object (unless you create the object yourself)

🔥 it 👍🏻 



> Nothing explicit is written about `BottleSpecification`, though the getter does appear to be missing from https://rubydoc.brew.sh/Formula (we don't mark it as private so this is probably actually a doc generation error but I suppose that means we can argue that it's been private all along).

Agreed on private here.

This makes me wonder if we want to do some wider "mark everything as explicitly public or private" effort.

**Comment by Bo98** [comment]
> This makes me wonder if we want to do some wider "mark everything as explicitly public or private" effort.

Probably. If it's not explicitly marked private then we shouldn't be surprised if people use it.

When it's private, the rubydocs page has a big warning. If it isn't then it appears as normal.

**Comment by Bo98** [comment]
Ok I've deleted everything that was nominated for deprecation.

**Comment by MikeMcQuaid** [comment]
> When it's private, the rubydocs page has a big warning. If it isn't then it appears as normal.

On the flip side there is this: https://rubydoc.brew.sh: 

> Probably. If it's not explicitly marked private then we shouldn't be surprised if people use it.

**Review by MikeMcQuaid — APPROVED** [review]
:shipit:

**Comment by Bo98** [comment]
> and classes linked from there

This can extend quite far. There's a link from Formula to SoftwareSpec to BottleSpecification.

**Pull request merged**